Unify SMTP mail sending

GutoVeronezi opened the pull request:
### Description
ACS has several features that need to send email through SMTP; However the code to send emails is repeated along the features.    
Code repetition does not guarantee that sending email will work the same way, makes maintenance difficult and multiplies the work to add a new option, like `StartTLS`.

This PR intends to create an util to send email through SMTP, to minimize code and to facilite maintenance, unit tests creation and troubleshooting.

### Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale
- [ ] Major
- [x] Minor

### How Has This Been Tested?
I configured SMTP global configs to project, alert and quota.
- To quota I added 10 credits to my user and set 20 as min balance, then I called `quota update` API via Cloudmonkey;
- To project I added an user;
- To alert called `generate alert` API via Cloudmonkey;

DaanHoogland commented:
nice initiative @GutoVeronezi, thanks. I'm starting smoke tests and will review when done.

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_check_mark: centos8 :heavy_check_mark: debian. SL-JID 497

DaanHoogland commented:
@blueorangutan test matrix

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins matrix job (centos7 mgmt + xs71, centos7 mgmt + vmware65, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-571)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33341 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t571-kvm-centos7.zip
Smoke tests completed. 88 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-570)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33784 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t570-xenserver-71.zip
Smoke tests completed. 88 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-572)</b>
Environment: vmware-65u2 (x2), Advanced Networking with Mgmt server 7
Total time taken: 41826 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t572-vmware-65u2.zip
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_volumes.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 86 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_10_list_volumes | `Failure` | 366.55 | test_volumes.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 354.13 | test_vpc_vpn.py

DaanHoogland reviewed (approved):
clgtm, smoke test failures seem not related to this change/travis passes. good cleanup.

GabrielBrascher commented on utils/src/test/java/org/apache/cloudstack/utils/mailing/SMTPMailSenderTest.java:
```diff
@@ -0,0 +1,618 @@
+/*
+ * Copyright 2021 The Apache Software Foundation.
```
I did not pay attention to these source code headers with Copyright but I now realize that you have been using this pattern on your PRs. Can you please change them all to fit the [Source File Headers for Code Developed at the ASF](https://www.apache.org/legal/src-headers.html#headers)?

Each source file should include the following license header -- **note that there should be no copyright notice in the header**:
```
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
```

Apache also makes it is very clear that such protocol (not adding Copyright) must be applied to Apache source code (e.g. CloudStack) but it does not apply to developers outside the ASF who are applying the Apache License to their work.

DaanHoogland replied:
ah, never thought it was a problem, we have loads of files with notices like this. Do we need a clean up action @GabrielBrascher ?

GabrielBrascher replied:
@DaanHoogland that's a good question.

This header difference is a bit confusing especially due to the fact that Apache license 2.0 adds the Copyright, but at the same time, the documentation regarding Apache's source code do say to **not add** the copyright.

I think that @GutoVeronezi is used to work with https://www.apache.org/licenses/LICENSE-2.0 and then added it on his PRs. I have nothing against it; however, Apache is very clear about how to proceed in terms of [Source File Headers for Code Developed at the ASF](https://www.apache.org/legal/src-headers.html#headers), which I think fits on any CloudStack source header, regardless of who donated the contribution to the Apache

GabrielBrascher replied:
Also, Apache refers to that document as:

> This document describes how Apache committers and PMC members should handle source file licensing and copyright notices.
It does not apply to developers outside the ASF who are applying the Apache License to their work.

The code is merged always by a **committer** that are adding a **contributor** code. Therefore, my understanding is that either @GutoVeronezi changes his code or then he leaves as it is but on the merge process the committer/PMC doing it should follow the Apache legal guidelines by adding a commit removing any Copyright on headers.

GutoVeronezi replied:
@GabrielBrascher I changed my projects, so from now on they will contain ASF's header. Do I need to change all my open PRs to change it or leave them to be changed when they are merged?

GabrielBrascher replied:
Thanks for the update @GutoVeronezi.

If possible to change them it would be nice, otherewise I would ask to add an observation regarding the source code header (if there is one of the PRs still open that we have not discussed the header yet).

One easy way of updating it (via UI) is with review, anyone (including you) can add code changes via reviews and them can be added to the PRs commits.

I will propose such changes, if you want you can commit them or ignore.

GabrielBrascher reviewed (commented):
@GutoVeronezi as discussed, here follows some suggestions for changing the headers.

GutoVeronezi commented:
@GabrielBrascher done.

GabrielBrascher reviewed (approved):
Code LGTM

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_check_mark: centos8 :heavy_check_mark: debian. SL-JID 539

blueorangutan commented:
<b>Trillian test result (tid-632)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 44370 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t632-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Smoke tests completed. 87 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_07_deploy_kubernetes_ha_cluster | `Failure` | 3606.10 | test_kubernetes_clusters.py
test_08_deploy_and_upgrade_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 40.90 | test_kubernetes_clusters.py

blueorangutan commented:
<b>Trillian test result (tid-641)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 47915 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t641-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_iso.py
Intermittent failure detected: /marvin/tests/smoke/test_loadbalance.py
Intermittent failure detected: /marvin/tests/smoke/test_network.py
Intermittent failure detected: /marvin/tests/smoke/test_pvlan.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 85 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_create_pvlan_network | `Error` | 0.04 | test_pvlan.py
test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false | `Failure` | 331.64 | test_routers_network_ops.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 457.73 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Failure` | 500.28 | test_vpc_redundant.py

GutoVeronezi commented:
@DaanHoogland @GabrielBrascher is there anything else to do?

rhtyd commented:
@GutoVeronezi do you have test results from testing the changes? For example setup the smtp settings and see if it actually works, share tests with screenshots etc.
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

GutoVeronezi commented:
@rhtyd 

As mentioned before, I set SMTP global settings to project, alert and quota and enabled StartTLS (global setting); Then I allowed `Less secure app access` on my Google account and set the following settings:
```
server address: smtp.gmail.com
username: <gmail address>
password: <gmail password>
port (TLS): 587
```

- To alert, I started mgmt;
- To quota, I added 10 credits to my user and set 20 as min balance, then I called API `quota update` via Cloudmonkey;
- To project, I enabled global setting `project.invite.required` and added an user via email;

Next the screenshot of received emails:

![image](https://user-images.githubusercontent.com/38945620/119571840-3d810680-bd88-11eb-81f1-f0a5e3613846.png)

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-752)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 100576 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t752-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_accounts.py
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_async_job.py
Intermittent failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermittent failure detected: /marvin/tests/smoke/test_deploy_vm_iso.py
Intermittent failure detected: /marvin/tests/smoke/test_deploy_vms_with_varied_deploymentplanners.py
Intermittent failure detected: /marvin/tests/smoke/test_deploy_vm_with_userdata.py
Intermittent failure detected: /marvin/tests/smoke/test_diagnostics.py
Intermittent failure detected: /marvin/tests/smoke/test_domain_service_offerings.py
Intermittent failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermittent failure detected: /marvin/tests/smoke/test_iso.py
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_supported_versions.py
Intermittent failure detected: /marvin/tests/smoke/test_list_ids_parameter.py
Intermittent failure detected: /marvin/tests/smoke/test_loadbalance.py
Intermittent failure detected: /marvin/tests/smoke/test_metrics_api.py
Intermittent failure detected: /marvin/tests/smoke/test_multipleips_per_nic.py
Intermittent failure detected: /marvin/tests/smoke/test_nested_virtualization.py
Intermittent failure detected: /marvin/tests/smoke/test_network_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_network.py
Intermittent failure detected: /marvin/tests/smoke/test_nic_adapter_type.py
Intermittent failure detected: /marvin/tests/smoke/test_password_server.py
Intermittent failure detected: /marvin/tests/smoke/test_portforwardingrules.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 63 look OK, 25 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestAccounts>:setup | `Error` | 0.00 | test_accounts.py
ContextSuite context=TestAddVmToSubDomain>:setup | `Error` | 0.00 | test_accounts.py
test_DeleteDomain | `Error` | 3.99 | test_accounts.py
test_forceDeleteDomain | `Failure` | 3.91 | test_accounts.py
ContextSuite context=TestRemoveUserFromAccount>:setup | `Error` | 4.95 | test_accounts.py
ContextSuite context=TestDomainsServiceOfferings>:setup | `Error` | 1559.06 | test_domain_service_offerings.py
ContextSuite context=TestInternalLb>:setup | `Error` | 0.00 | test_internal_lb.py
ContextSuite context=TestDeployVmWithAffinityGroup>:setup | `Error` | 0.00 | test_affinity_groups_projects.py
test_01_1_create_iso_with_checksum_sha1_negative | `Error` | 66.36 | test_iso.py
test_01_create_iso_with_checksum_sha1 | `Error` | 66.35 | test_iso.py
test_02_create_iso_with_checksum_sha256 | `Error` | 66.42 | test_iso.py
test_03_create_iso_with_checksum_md5 | `Error` | 66.36 | test_iso.py
test_04_create_iso_with_no_checksum | `Error` | 66.37 | test_iso.py
test_01_create_iso | `Failure` | 1511.66 | test_iso.py
ContextSuite context=TestISO>:setup | `Error` | 3022.68 | test_iso.py
ContextSuite context=TestAsyncJob>:setup | `Error` | 0.00 | test_async_job.py
ContextSuite context=TestLoadBalance>:setup | `Error` | 0.00 | test_loadbalance.py
ContextSuite context=TestDeployVirtioSCSIVM>:setup | `Error` | 0.00 | test_deploy_virtio_scsi_vm.py
test_list_clusters_metrics | `Error` | 1512.52 | test_metrics_api.py
test_list_vms_metrics | `Error` | 0.13 | test_metrics_api.py
ContextSuite context=TestDeployVMFromISO>:setup | `Error` | 0.00 | test_deploy_vm_iso.py
ContextSuite context=TestDeployVmWithVariedPlanners>:setup | `Error` | 0.00 | test_deploy_vms_with_varied_deploymentplanners.py
ContextSuite context=TestNetworkACL>:setup | `Error` | 0.00 | test_network_acl.py
ContextSuite context=TestDeployVmWithUserData>:setup | `Error` | 0.00 | test_deploy_vm_with_userdata.py
ContextSuite context=TestRemoteDiagnostics>:setup | `Error` | 0.00 | test_diagnostics.py
test_delete_account | `Error` | 1513.74 | test_network.py
test_delete_network_while_vm_on_it | `Error` | 1.10 | test_network.py
test_deploy_vm_l2network | `Error` | 1.10 | test_network.py
test_l2network_restart | `Error` | 2.19 | test_network.py
ContextSuite context=TestPortForwarding>:setup | `Error` | 3.34 | test_network.py
ContextSuite context=TestPublicIP>:setup | `Error` | 0.98 | test_network.py
test_reboot_router | `Failure` | 0.05 | test_network.py
test_releaseIP | `Error` | 0.41 | test_network.py
ContextSuite context=TestRouterRules>:setup | `Error` | 0.49 | test_network.py
ContextSuite context=TestAdapterTypeForNic>:setup | `Error` | 0.00 | test_nic_adapter_type.py
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_02_deploy_and_upgrade_kubernetes_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_04_basic_lifecycle_kubernetes_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_05_delete_kubernetes_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_06_deploy_invalid_kubernetes_ha_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_08_deploy_and_upgrade_kubernetes_ha_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 0.00 | test_kubernetes_clusters.py
test_01_add_delete_kubernetes_supported_version | `Error` | 60.60 | test_kubernetes_supported_versions.py
ContextSuite context=TestListIdsParams>:setup | `Error` | 0.00 | test_list_ids_parameter.py
test_nic_secondaryip_add_remove | `Error` | 1512.41 | test_multipleips_per_nic.py
ContextSuite context=TestNestedVirtualization>:setup | `Error` | 0.00 | test_nested_virtualization.py
ContextSuite context=TestIsolatedNetworksPasswdServer>:setup | `Error` | 0.00 | test_password_server.py
ContextSuite context=TestPortForwardingRules>:setup | `Error` | 0.00 | test_portforwardingrules.py
ContextSuite context=TestPrivateGwACL>:setup | `Error` | 0.00 | test_privategw_acl.py
ContextSuite context=TestProjectSuspendActivate>:setup | `Error` | 1518.95 | test_projects.py

blueorangutan commented:
<b>Trillian test result (tid-775)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 34383 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4954-t775-kvm-centos7.zip
Smoke tests completed. 88 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
I'm going to trust this one and regard the testfailures earlier as environmental.

@rhtyd satisfied with @GutoVeronezi 's answers?

GutoVeronezi commented:
@DaanHoogland @GabrielBrascher @rhtyd  is there anything else to do?

DaanHoogland commented:
@rhtyd want to do the honours, please?

rhtyd commented:
thanks lgtm, @GutoVeronezi can you fix the conflict though?
@DaanHoogland you may merge this if/when Travis passes.

GutoVeronezi commented:
@rhtyd @DaanHoogland conflicts solved, waiting for Travis results.

GutoVeronezi commented:
@DaanHoogland all Travis checks have passed.
